<launch>
    
    <node name="base_to_aruco" pkg="tf2_ros" type="static_transform_publisher" args="0 0 0 0 0 0 base aruco/body" />
    
	<group ns="zero">
	    <!--                 -->
	    <!-- Pose estimation -->
	    <!--                 -->
	    <param name="aruco_description" textfile="$(find aruco_pose_estimation)/urdf/aruco_v2.urdf" />

	    <include file="$(find aruco_recognition)/launch/samples/recognition_v3_cmyk.launch">
	    	<arg name="image_top" value="image_raw"/>
	    </include>    	

	    <node name="pose_estimator" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID0"/> # target AR marker's ID
		<param name="marker_frame_id" value="aruco/body"/>
		<param name="camera_frame_id" value="SSP0_camera"/>
		<param name="marker_length" value="0.25"/>
	    </node>
	    
	    <node name="pose_estimator_robot1" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID1"/> # target AR marker's ID
		<param name="marker_frame_id" value="SSP0_camera"/>
		<param name="camera_frame_id" value="SSP0_robot1"/>
		<param name="marker_length" value="0.20"/>
	    </node>

		<node name="pose_estimator_robot2" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID2"/> # target AR marker's ID
		<param name="marker_frame_id" value="SSP0_camera"/>
		<param name="camera_frame_id" value="SSP0_robot2"/>
		<param name="marker_length" value="0.20"/>
	    </node>
    </group>
                                    
    <group ns="one">
	    <!--                 -->
	    <!-- Pose estimation -->
	    <!--                 -->
	    <param name="aruco_description" textfile="$(find aruco_pose_estimation)/urdf/aruco_v2.urdf" />

	    <include file="$(find aruco_recognition)/launch/samples/recognition_v3_cmyk.launch">
	    	<arg name="image_top" value="image_raw"/>
	    </include>    	

	    <node name="pose_estimator" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID0"/> # target AR marker's ID
		<param name="marker_frame_id" value="aruco/body"/>
		<param name="camera_frame_id" value="SSP1_camera"/>
		<param name="marker_length" value="0.25"/>
	    </node>
	    
	    <node name="pose_estimator_robot1" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID1"/> # target AR marker's ID
		<param name="marker_frame_id" value="SSP1_camera"/>
		<param name="camera_frame_id" value="SSP1_robot1"/>
		<param name="marker_length" value="0.20"/>
	    </node>

		<node name="pose_estimator_robot2" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID2"/> # target AR marker's ID
		<param name="marker_frame_id" value="SSP1_camera"/>
		<param name="camera_frame_id" value="SSP1_robot2"/>
		<param name="marker_length" value="0.20"/>
	    </node>
    </group>
    
    <group ns="two">
	    <!--                 -->
	    <!-- Pose estimation -->
	    <!--                 -->
	    <param name="aruco_description" textfile="$(find aruco_pose_estimation)/urdf/aruco_v2.urdf" />

	    <include file="$(find aruco_recognition)/launch/samples/recognition_v3_cmyk.launch">
	    	<arg name="image_top" value="image_raw"/>
	    </include>    	

	    <node name="pose_estimator" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID0"/> # target AR marker's ID
		<param name="marker_frame_id" value="aruco/body"/>
		<param name="camera_frame_id" value="SSP2_camera"/>
		<param name="marker_length" value="0.25"/>
	    </node>
	    
	    <node name="pose_estimator_robot1" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID1"/> # target AR marker's ID
		<param name="marker_frame_id" value="SSP2_camera"/>
		<param name="camera_frame_id" value="SSP2_robot1"/>
		<param name="marker_length" value="0.20"/>
	    </node>

		<node name="pose_estimator_robot2" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID2"/> # target AR marker's ID
		<param name="marker_frame_id" value="SSP2_camera"/>
		<param name="camera_frame_id" value="SSP2_robot2"/>
		<param name="marker_length" value="0.20"/>
	    </node>
    </group>

	<group ns="three">
	    <!--                 -->
	    <!-- Pose estimation -->
	    <!--                 -->
	    <param name="aruco_description" textfile="$(find aruco_pose_estimation)/urdf/aruco_v2.urdf" />

	    <include file="$(find aruco_recognition)/launch/samples/recognition_v3_cmyk.launch">
	    	<arg name="image_top" value="image_raw"/>
	    </include>    	

	    <node name="pose_estimator" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID0"/> # target AR marker's ID
		<param name="marker_frame_id" value="aruco/body"/>
		<param name="camera_frame_id" value="SSP3_camera"/>
		<param name="marker_length" value="0.25"/>
	    </node>
	    
	    <node name="pose_estimator_robot1" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID1"/> # target AR marker's ID
		<param name="marker_frame_id" value="SSP3_camera"/>
		<param name="camera_frame_id" value="SSP3_robot1"/>
		<param name="marker_length" value="0.20"/>
	    </node>

		<node name="pose_estimator_robot2" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID2"/> # target AR marker's ID
		<param name="marker_frame_id" value="SSP3_camera"/>
		<param name="camera_frame_id" value="SSP3_robot2"/>
		<param name="marker_length" value="0.20"/>
	    </node>
    </group>

	<group ns="four">
	    <!--                 -->
	    <!-- Pose estimation -->
	    <!--                 -->
	    <param name="aruco_description" textfile="$(find aruco_pose_estimation)/urdf/aruco_v2.urdf" />

	    <include file="$(find aruco_recognition)/launch/samples/recognition_v3_cmyk.launch">
	    	<arg name="image_top" value="image_raw"/>
	    </include>    	

	    <node name="pose_estimator" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID0"/> # target AR marker's ID
		<param name="marker_frame_id" value="aruco/body"/>
		<param name="camera_frame_id" value="SSP4_camera"/>
		<param name="marker_length" value="0.25"/>
	    </node>
	    
	    <node name="pose_estimator_robot1" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID1"/> # target AR marker's ID
		<param name="marker_frame_id" value="SSP4_camera"/>
		<param name="camera_frame_id" value="SSP4_robot1"/>
		<param name="marker_length" value="0.20"/>
	    </node>

		<node name="pose_estimator_robot2" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID2"/> # target AR marker's ID
		<param name="marker_frame_id" value="SSP4_camera"/>
		<param name="camera_frame_id" value="SSP4_robot2"/>
		<param name="marker_length" value="0.20"/>
	    </node>
    </group>

<group ns="five">
	    <!--                 -->
	    <!-- Pose estimation -->
	    <!--                 -->
	    <param name="aruco_description" textfile="$(find aruco_pose_estimation)/urdf/aruco_v2.urdf" />

	    <include file="$(find aruco_recognition)/launch/samples/recognition_v3_cmyk.launch">
	    	<arg name="image_top" value="image_raw"/>
	    </include>    	

	    <node name="pose_estimator" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID0"/> # target AR marker's ID
		<param name="marker_frame_id" value="aruco/body"/>
		<param name="camera_frame_id" value="SSP5_camera"/>
		<param name="marker_length" value="0.25"/>
	    </node>
	    
	    <node name="pose_estimator_robot1" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID1"/> # target AR marker's ID
		<param name="marker_frame_id" value="SSP5_camera"/>
		<param name="camera_frame_id" value="SSP5_robot1"/>
		<param name="marker_length" value="0.20"/>
	    </node>

		<node name="pose_estimator_robot2" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID2"/> # target AR marker's ID
		<param name="marker_frame_id" value="SSP5_camera"/>
		<param name="camera_frame_id" value="SSP5_robot2"/>
		<param name="marker_length" value="0.20"/>
	    </node>
    </group>

	<group ns="six">
	    <!--                 -->
	    <!-- Pose estimation -->
	    <!--                 -->
	    <param name="aruco_description" textfile="$(find aruco_pose_estimation)/urdf/aruco_v2.urdf" />

	    <include file="$(find aruco_recognition)/launch/samples/recognition_v3_cmyk.launch">
	    	<arg name="image_top" value="image_raw"/>
	    </include>    	

	    <node name="pose_estimator" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID0"/> # target AR marker's ID
		<param name="marker_frame_id" value="aruco/body"/>
		<param name="camera_frame_id" value="SSP6_camera"/>
		<param name="marker_length" value="0.25"/>
	    </node>
	    
	    <node name="pose_estimator_robot1" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID1"/> # target AR marker's ID
		<param name="marker_frame_id" value="SSP6_camera"/>
		<param name="camera_frame_id" value="SSP6_robot1"/>
		<param name="marker_length" value="0.20"/>
	    </node>

		<node name="pose_estimator_robot2" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID2"/> # target AR marker's ID
		<param name="marker_frame_id" value="SSP6_camera"/>
		<param name="camera_frame_id" value="SSP6_robot2"/>
		<param name="marker_length" value="0.20"/>
	    </node>
    </group>

	<group ns="seven">
	    <!--                 -->
	    <!-- Pose estimation -->
	    <!--                 -->
	    <param name="aruco_description" textfile="$(find aruco_pose_estimation)/urdf/aruco_v2.urdf" />

	    <include file="$(find aruco_recognition)/launch/samples/recognition_v3_cmyk.launch">
	    	<arg name="image_top" value="image_raw"/>
	    </include>    	

	    <node name="pose_estimator" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID0"/> # target AR marker's ID
		<param name="marker_frame_id" value="aruco/body"/>
		<param name="camera_frame_id" value="SSP7_camera"/>
		<param name="marker_length" value="0.25"/>
	    </node>
	    
	    <node name="pose_estimator_robot1" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID1"/> # target AR marker's ID
		<param name="marker_frame_id" value="SSP7_camera"/>
		<param name="camera_frame_id" value="SSP7_robot1"/>
		<param name="marker_length" value="0.20"/>
	    </node>

		<node name="pose_estimator_robot2" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID2"/> # target AR marker's ID
		<param name="marker_frame_id" value="SSP7camera"/>
		<param name="camera_frame_id" value="SSP7robot2"/>
		<param name="marker_length" value="0.20"/>
	    </node>
    </group>

	<group ns="eight">
	    <!--                 -->
	    <!-- Pose estimation -->
	    <!--                 -->
	    <param name="aruco_description" textfile="$(find aruco_pose_estimation)/urdf/aruco_v2.urdf" />

	    <include file="$(find aruco_recognition)/launch/samples/recognition_v3_cmyk.launch">
	    	<arg name="image_top" value="image_raw"/>
	    </include>    	

	    <node name="pose_estimator" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID0"/> # target AR marker's ID
		<param name="marker_frame_id" value="aruco/body"/>
		<param name="camera_frame_id" value="SSP8_camera"/>
		<param name="marker_length" value="0.25"/>
	    </node>
	    
	    <node name="pose_estimator_robot1" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID1"/> # target AR marker's ID
		<param name="marker_frame_id" value="SSP8_camera"/>
		<param name="camera_frame_id" value="SSP8_robot1"/>
		<param name="marker_length" value="0.20"/>
	    </node>

		<node name="pose_estimator_robot2" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID2"/> # target AR marker's ID
		<param name="marker_frame_id" value="SSP8_camera"/>
		<param name="camera_frame_id" value="SSP8_robot2"/>
		<param name="marker_length" value="0.20"/>
	    </node>
    </group>

	<group ns="nine">
	    <!--                 -->
	    <!-- Pose estimation -->
	    <!--                 -->
	    <param name="aruco_description" textfile="$(find aruco_pose_estimation)/urdf/aruco_v2.urdf" />

	    <include file="$(find aruco_recognition)/launch/samples/recognition_v3_cmyk.launch">
	    	<arg name="image_top" value="image_raw"/>
	    </include>    	

	    <node name="pose_estimator" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID0"/> # target AR marker's ID
		<param name="marker_frame_id" value="aruco/body"/>
		<param name="camera_frame_id" value="SSP9_camera"/>
		<param name="marker_length" value="0.25"/>
	    </node>
	    
	    <node name="pose_estimator_robot1" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID1"/> # target AR marker's ID
		<param name="marker_frame_id" value="SSP9_camera"/>
		<param name="camera_frame_id" value="SSP9_robot1"/>
		<param name="marker_length" value="0.20"/>
	    </node>

		<node name="pose_estimator_robot2" pkg="nodelet" type="nodelet"
		  args="standalone aruco_pose_estimation/RelativePoseEstimator" output="screen">
		<remap from="barcode_data" to="qr_data" />
		<param name="cache_barcode" value="true" />
		<param name="target_name" value="ID2"/> # target AR marker's ID
		<param name="marker_frame_id" value="SSP9_camera"/>
		<param name="camera_frame_id" value="SSP9_robot2"/>
		<param name="marker_length" value="0.20"/>
	    </node>
    </group>


    <!--               -->
    <!-- Visualization -->
    <!--               -->
    <node name="pose_viewer" pkg="rviz" type="rviz"
          args="--display-config $(find aruco_pose_estimation)/rviz/online_pose_estimation_v4.rviz" output="screen">
		  <param name="/use_sim_time" value="true"/>
    </node>
    
</launch>
